### PR TITLE
[RDY] lua: vim.deepcopy and vim.tbl_extend use empty_dict() instead of {} for empty_dict()

### LINKED ISSUE #94: [rust_analyzer] initalizationOptions causing ra_lsp_server an error
Because of how vim `json_encode` transform vim data into JSON data, the lua value `{}` which is the default value for `init_options` is being encoded into an empty JSON array.
But `ra_lsp_server` expect to receive a JSON map for decoding the init options and emit an error: `failed to deserialize config: invalid length 0, expected struct ServerConfig with 12 elements` when receiving an empty initializationOptions by vim.lsp.
Hopefully rust-analyzer handle this error by just logging it and sticking with default values, but still the error message appears, which can be actually confusing.

### PR DESCRIPTION
**Pull request by h-michael**
fix: https://github.com/neovim/nvim-lsp/issues/94

cc: @justinmk

### CODE REVIEW

File: test/functional/lua/vim_spec.lua [hunk 1]
```diff
@@ -322,6 +322,48 @@ describe('lua stdlib', function()
...
+
+    local is_empty_dic = exec_lua([[
+      local a = vim.empty_dict()
+      local b = vim.deepcopy(a)
+
+      local count = 0
+      for _ in pairs(b) do count = count + 1 end
+
+      return getmetatable(b) == vim._empty_dict_mt
+        and count == 0
```
Comment (h-michael) [inline test/functional/lua/vim_spec.lua]: I want to add `and tostring(b) ~= tostring(a)`. But `tostring(vim.empty_dict()` always returns `vim.empty_dict()`.
So I can't compare table ids.

**Comment by h-michael** [comment]
I don't know how can I fix this.

https://travis-ci.org/neovim/neovim/jobs/649771817#L6954-L6959
```
[  ERROR   ] test/functional/terminal/tui_spec.lua @ 232: TUI accepts ASCII control sequences
/home/travis/build/neovim/neovim/runtime/lua/vim/shared.lua:25: attempt to call field 'empty_dict' (a nil value)

stack traceback:
	/home/travis/build/neovim/neovim/runtime/lua/vim/shared.lua:25: in function 'get_default_attr_ids'
	test/functional/terminal/tui_spec.lua:237: in function <test/functional/terminal/tui_spec.lua:232>
```

**Comment by h-michael** [comment]
I think that the reason why test cannot be executed well is related to this comment.
https://github.com/neovim/neovim/pull/11859#discussion_r378190779

File: runtime/lua/vim/shared.lua [hunk 2]
```diff
@@ -174,9 +174,19 @@ function vim.tbl_extend(behavior, ...)
   if (behavior ~= 'error' and behavior ~= 'keep' and behavior ~= 'force') then
     error('invalid "behavior": '..tostring(behavior))
   end
+
+  if select('#', ...) < 2 then
+    error('wrong number of arguments (given '..tostring(1 + select('#', ...))..', expected at least 3)')
+  end
+
   local ret = {}
+  if getmetatable(select(1, ...)) == vim._empty_dict_mt then
```
Comment (bfredl) [inline runtime/lua/vim/shared.lua]: this should be written as `if vim and getmetatable(select(1, ...)) == vim._empty_dict_mt then`

File: runtime/lua/vim/shared.lua [hunk 3]
```diff
@@ -20,6 +20,11 @@ vim.deepcopy = (function()
   local deepcopy_funcs = {
     table = function(orig)
       local copy = {}
+
+      if getmetatable(orig) == vim._empty_dict_mt then
```
Comment (bfredl) [inline runtime/lua/vim/shared.lua]: `if vim and getmetatable(orig) == vim._empty_dict_mt then`

Comment (bfredl) [reply runtime/lua/vim/shared.lua]: Sorry, actually `if vim._empty_dict_mt ~= nil and getmetatable(orig) == vim._empty_dict_mt then`

File: test/functional/lua/vim_spec.lua [hunk 4]
```diff
@@ -319,9 +319,43 @@ describe('lua stdlib', function()
...
+    ]]))
+
+    assert(exec_lua([[
+      local a = {x = vim.empty_dict(), y = {}}
+      local b = vim.deepcopy(a)
+
+      local count = 0
+      for _ in pairs(b) do count = count + 1 end
+
+      return getmetatable(b.x) == vim._empty_dict_mt
```
Comment (bfredl) [inline test/functional/lua/vim_spec.lua]: A more canonical way rather than checking `vim._empty_dict_mt` manually, is to use `vim.tbl_islist(b.x)`

**Comment by h-michael** [comment]
builds.sr.ht for FreeBSD failed.
Because of it is trying to download "cmake-3.16.3.txz", but it no longer exists.
Now there is "cmake-3.16.4.txz"

File: test/functional/lua/vim_spec.lua [hunk 5]
```diff
@@ -319,9 +319,39 @@ describe('lua stdlib', function()
 
       return b.x[1] == 1 and b.x[2] == 2 and b.y == 5 and count == 2
              and tostring(a) ~= tostring(b)
-    ]])
+    ]]))
+
+    assert(exec_lua([[
```
Comment (bfredl) [inline test/functional/lua/vim_spec.lua]: I think we normally use `ok()` for expecting `true` value

File: test/functional/lua/vim_spec.lua [hunk 6]
```diff
@@ -319,9 +320,39 @@ describe('lua stdlib', function()
...
       return b.x[1] == 1 and b.x[2] == 2 and b.y == 5 and count == 2
              and tostring(a) ~= tostring(b)
-    ]])
+    ]]))
+
+    ok(exec_lua([[
+      local a = {}
+      local b = vim.deepcopy(a)
+
+      local count = 0
```
Comment (bfredl) [inline test/functional/lua/vim_spec.lua]: We don't need to do it now but later we could add `vim.tbl_count()` or something.

**Review by bfredl — APPROVED** [review]

**Pull request merged**